src/go/k8s/internal/client/factory.go

```diff
@@ -0,0 +1,183 @@
+)
+
+// ClientFactory is responsible for creating both high-level and low-level clients used in our
+// controllers.
+//
+// Calling its `Kafka*` methods will initialize a low-level kgo.Client instance
+// based on the connection parameters contained within the corresponding CRD struct passed in
+// at method invocation.
+//
+// Calling its `RedpandaAdmin*` methods will initialize a low-level rpadmin.AdminAPI instance
+// based on the connection parameters contained within the corresponding CRD struct passed in
+// at method invocation.
+//
+// Additionally, high-level clients that act as functional wrappers to the low-level
+// clients to ease high-level operations (i.e. syncing all primitives for a v1alpha2.User) may
+// be initialized through calling the appropriate `*Client` methods.
+type ClientFactory interface {
+	// WithDialer causes the underlying connections to use a different dialer implementation.
+	WithDialer(dialer redpanda.DialContextFunc) ClientFactory
+	// WithLogger specifies a logger to be passed to the underlying clients.
+	WithLogger(logger logr.Logger) ClientFactory
+
+	// KafkaClient initializes a kgo.Client based on the spec of the passed in struct.
+	// The struct *must* implement either the v1alpha2.KafkaConnectedObject interface of the v1alpha2.ClusterReferencingObject
+	// interface to properly initialize.
+	KafkaClient(ctx context.Context, object client.Object, opts ...kgo.Opt) (*kgo.Client, error)
+
+	// RedpandaAdminClient initializes a rpadmin.AdminAPI client based on the spec of the passed in struct.
+	// The struct *must* implement either the v1alpha2.AdminConnectedObject interface of the v1alpha2.ClusterReferencingObject
+	// interface to properly initialize.
+	RedpandaAdminClient(ctx context.Context, object client.Object) (*rpadmin.AdminAPI, error)
+}
+
+type clientFactory struct {
+	client.Client
+	logger logr.Logger
+	config *rest.Config
+	dialer redpanda.DialContextFunc
+}
+
+var _ ClientFactory = (*clientFactory)(nil)
+
+func NewClientFactory(config *rest.Config) (ClientFactory, error) {
+	s := runtime.NewScheme()
+	if err := clientgoscheme.AddToScheme(s); err != nil {
+		return nil, err
+	}
+
+	if err := redpandav1alpha2.AddToScheme(s); err != nil {
+		return nil, err
+	}
+
+	if err := redpandav1alpha1.AddToScheme(s); err != nil {
+		return nil, err
+	}
+
+	c, err := client.New(config, client.Options{Scheme: s})
+	if err != nil {
+		return nil, err
+	}
+
+	return &clientFactory{
+		Client: c,
+		config: config,
+		logger: logr.Discard(),
+	}, nil
+}
+
+func (c *clientFactory) WithDialer(dialer redpanda.DialContextFunc) ClientFactory {
+	c.dialer = dialer
+	return c
+}
+
+func (c *clientFactory) WithLogger(logger logr.Logger) ClientFactory {
+	c.logger = logger
+	return c
+}
+
+func (c *clientFactory) KafkaClient(ctx context.Context, obj client.Object, opts ...kgo.Opt) (*kgo.Client, error) {
+	// if we pass in a Redpanda cluster, just use it
```

Comment:
nit: I feel a single type switch here would be a lot more understandable. Is there a reason that's not suitable? (Could a single type implement multiple of these interfaces?)

Reply:
For me the readability comes in trying to flatten some of the conditional nesting around some labeled helprs and the removal of having to use `fallthrough` handling coupled with an additional type assertion guard and case ordering in a type switch.

Here's this function's body roughly inlined with a switch:

```
switch o := obj.(type) {
case *redpandav1alpha2.Redpanda:
  return c.kafkaForCluster(ctx, o, opts...)

case redpandav1alpha2.ClusterReferencingObject:

  if ref := o.GetClusterRef(); ref != nil {
    var cluster redpandav1alpha2.Redpanda

    if err := c.Get(ctx, types.NamespacedName{Namespace: obj.GetNamespace(), Name: ref.Name}, &cluster); err != nil {
      return nil, err
    }

    return c.kafkaForCluster(ctx, &cluster, opts...)
  }
  
  if _, ok := obj.(redpandav1alpha2.ClusterReferencingObject); !ok {
    return nil, ErrInvalidKafkaClientObject
  }

  fallthrough

case redpandav1alpha2.KafkaConnectedObject:
  if spec := o.GetKafkaAPISpec(); spec != nil {
    var metricsNamespace *string

    metricsNamespace = nil
    if ons, ok := obj.(redpandav1alpha2.KafkaConnectedObjectWithMetrics); ok {
      metricsNamespace = ons.GetMetricsNamespace()
    }

    return c.kafkaForSpec(ctx, obj.GetNamespace(), metricsNamespace, spec, opts...)
  }
}

return nil, ErrInvalidKafkaClientObject
```

Reply:
And, wow, totally learned something I didn't know previously -- `fallthrough` doesn't work on type switches (guess it makes sense since there's also some casting going on and you can potentially violate the type-check if `fallthrough` was allowed).

So, I don't think the switch would work in general.

Reply:
aaaaaaaaaaaaaaah, I gotcha. The `fallthrough` when something is `nil` wasn't clear to me but that would have been my guess as to why a type switch wasn't used.

LGTM as is! As an option for the future:

The abstractions might pan out a bit more smoothly if the `client` package has an intermediate representation that `ClusterRef` and `KafkaAPISpec` know how to produce. We could then slap an interface onto `Topic` and `User` that handles the change of `if t.ClusterRef != nil { return t.ClusterRef.ConnectionSpec(client)}`. That would resolve some of the strangeness of the CRDs being imported by the client package as well. Though we'd have to move the client package out of `internal`, which I'm quite okay with.